### PR TITLE
[filebeat][input] Azure blob storage [refactor]

### PR DESCRIPTION
**Pull request by ShourieG**
# Type of change
## What does this PR do?

This is a complete refactor of the azure blob storage input module. Code style & structural changes have been made to make the code more organised, compact & readable.

## Why is it important?
This PR keeps the code up to modern go standards, reducing unnecessary complexities and increasing code maintainability.

## Checklist
- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
~~- [ ] I have made corresponding change to the default configuration files~~
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `input-azure-blob-storage.asciidoc` .

## Author's Checklist
1. Documentation.
2. Config format.
3. Code errors.
4. Design choices.

## Related issues
- Relates #32438

### CODE REVIEW

File: x-pack/filebeat/docs/inputs/input-azure-blob-storage.asciidoc [hunk 1]
```diff
@@ -97,29 +97,12 @@ processing a certain number of files , decided by the `maximum available workers
...
     "azure": {
-        "blob": {
+        "storage": {
             "container": {
                 "name": "beatscontainer"
             },
-            "object": {
+            "blob": {
                 "content_type": "application/json",
-                "data": [
```
Comment (efd6) [inline x-pack/filebeat/docs/inputs/input-azure-blob-storage.asciidoc:106]: Why this change?

File: x-pack/filebeat/docs/inputs/input-azure-blob-storage.asciidoc [hunk 2]
```diff
@@ -130,16 +113,15 @@ processing a certain number of files , decided by the `maximum available workers
...
     
 *Some of the key attributes are as follows :-* 
 
     1. *message* : Original stringified blob data.
     2. *log.file.path* : Path of the blob in azure cloud.
     3. *azure.blob.container.name* : Name of the container from which the file has been read.
     4. *azure.blob.object.name* : Name of the file/blob which has been read.
-    5. *azure.blob.object.content_type* : Content type of the file/blob. You can find the supported content types <<supported-types,here>> .
-    6. *azure.blob.object.data* :  Objectified blob data , representing the contents of the file.
+    5. *azure.blob.object.content_type* : Content type of the file/blob. You can find the supported content types <<supported-types,here>>.
```
Comment (efd6) [inline x-pack/filebeat/docs/inputs/input-azure-blob-storage.asciidoc:124]: `azure.storage.blob.name` and `azure.storage.blob.content_type`?

File: x-pack/filebeat/input/azureblobstorage/client.go [hunk 3]
```diff
@@ -39,31 +35,31 @@ func fetchServiceClientWithSharedKeyCreds(url string, accountName string, cfg *s
...
-		return fetchBlobClientWithSharedKey(url, credential.ServiceCreds.SharedKeyCreds, log)
-	case types.ConnectionStringType:
-		return fetchBlobClientWithConnectionString(credential.ServiceCreds.ConnectionStrCreds, credential.ContainerName, credential.BlobName, log)
+	switch credential.serviceCreds.cType {
+	case sharedKeyType:
+		return fetchBlobClientWithSharedKey(url, credential.serviceCreds.sharedKeyCreds, log)
+	case connectionStringType:
+		return fetchBlobClientWithConnectionString(credential.serviceCreds.connectionStrCreds, credential.containerName, credential.blobName, log)
 	default:
 		return nil, fmt.Errorf("no valid service credential 'type' found")
```
Comment (efd6) [inline x-pack/filebeat/input/azureblobstorage/client.go:64]:
```suggestion
		return nil, fmt.Errorf("no valid service credential 'type' found: %s", credential.serviceCreds.cType)
```

File: x-pack/filebeat/input/azureblobstorage/input_test.go [hunk 4]
```diff
@@ -0,0 +1,328 @@
...
+	v2 "github.com/elastic/beats/v7/filebeat/input/v2"
+	beattest "github.com/elastic/beats/v7/libbeat/publisher/testing"
+	"github.com/elastic/beats/v7/x-pack/filebeat/input/azureblobstorage/mock"
+	conf "github.com/elastic/elastic-agent-libs/config"
+	"github.com/elastic/elastic-agent-libs/logp"
+)
+
+const (
+	beatsContainer   = "beatscontainer"
+	beatsContainer_2 = "beatscontainer2"
```
Comment (efd6) [inline x-pack/filebeat/input/azureblobstorage/input_test.go:28]:
```suggestion
	beatsContainer2 = "beatscontainer2"
```

File: x-pack/filebeat/input/azureblobstorage/input_test.go [hunk 5]
```diff
@@ -0,0 +1,328 @@
...
+	tests := []struct {
+		name            string
+		baseConfig      map[string]interface{}
+		mockHandler     func() http.Handler
+		expected        map[string]bool
+		isError         error
+		unexpectedError error
+	}{
+		{
+			name: "Test1_SingleContainerWithPoll_NoErr",
```
Comment (efd6) [inline x-pack/filebeat/input/azureblobstorage/input_test.go:41]: I don't think the `Testn_` prefix is needed.

File: x-pack/filebeat/input/azureblobstorage/input_test.go [hunk 6]
```diff
@@ -0,0 +1,328 @@
...
+				return input.Run(ctx, chanClient)
+			})
+
+			var timeout *time.Timer
+			if conf.PollInterval != nil {
+				timeout = time.NewTimer(1*time.Second + *conf.PollInterval)
+			} else {
+				timeout = time.NewTimer(5 * time.Second)
+			}
+			t.Cleanup(func() { _ = timeout.Stop() })
```
Comment (efd6) [inline x-pack/filebeat/input/azureblobstorage/input_test.go:274]:
```suggestion
			t.Cleanup(func() { timeout.Stop() })
```

File: x-pack/filebeat/input/azureblobstorage/job.go [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+func (j *job) timestamp() *time.Time {
+	return j.blob.Properties.LastModified
+}
+
+func (j *job) extractData(ctx context.Context) (*bytes.Buffer, error) {
+	var err error
+
+	get, err := j.client.Download(ctx, nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to download data from blob with error : %w", err)
```
Comment (efd6) [inline x-pack/filebeat/input/azureblobstorage/job.go:105]:
```suggestion
		return nil, fmt.Errorf("failed to download data from blob with error: %w", err)
```

File: x-pack/filebeat/input/azureblobstorage/job.go [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+		if err != nil {
+			j.log.Errorf(jobErrString, id, err)
+			return
+		}
+
+		reader := io.NopCloser(bytes.NewReader(data.Bytes()))
+		defer func() {
+			err = reader.Close()
+			if err != nil {
+				j.log.Errorf("failed to close io reader with error : %w", err)
```
Comment (efd6) [inline x-pack/filebeat/input/azureblobstorage/job.go:63]:
```suggestion
				j.log.Errorf("failed to close io reader with error: %w", err)
```

File: x-pack/filebeat/input/azureblobstorage/job.go [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+	defer func() {
+		err = reader.Close()
+		if err != nil {
+			err = fmt.Errorf("failed to close blob reader with error : %w", err)
+		}
+	}()
+
+	_, err = downloadedData.ReadFrom(reader)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read data from blob with error : %w", err)
```
Comment (efd6) [inline x-pack/filebeat/input/azureblobstorage/job.go:119]:
```suggestion
		return nil, fmt.Errorf("failed to read data from blob with error: %w", err)
```

File: x-pack/filebeat/input/azureblobstorage/mock/data.go [hunk 10]
```diff
@@ -0,0 +1,238 @@
...
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package mock
+
+import "errors"
+
+const (
+	beatsContainer   = "beatscontainer"
+	beatsContainer_2 = "beatscontainer2"
```
Comment (efd6) [inline x-pack/filebeat/input/azureblobstorage/mock/data.go:11]:
```suggestion
	beatsContainer2 = "beatscontainer2"
```

File: x-pack/filebeat/input/azureblobstorage/mock/mock.go [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+// you may not use this file except in compliance with the Elastic License.
+
+package mock
+
+import (
+	"net/http"
+	"strings"
+)
+
+//nolint:errcheck // We can ignore as this is just for testing
```
Comment (efd6) [inline x-pack/filebeat/input/azureblobstorage/mock/mock.go:12]: I'd prefer that this based the claim on the non-handleablility of response writer errors rather than it being testing.

File: x-pack/filebeat/docs/inputs/input-azure-blob-storage.asciidoc (hunk 1 above)
Comment (ShourieG) [reply x-pack/filebeat/docs/inputs/input-azure-blob-storage.asciidoc:106]: This was done in order to make the output follow the same structure as the aws s3 input. After a bit of thought the data field was seeming to be kind of useless as it was just the json parsed version of the data in the message field. I think the message field is sufficient here, it contains the original stringified data.

File: x-pack/filebeat/docs/inputs/input-azure-blob-storage.asciidoc (hunk 2 above)
Comment (ShourieG) [reply x-pack/filebeat/docs/inputs/input-azure-blob-storage.asciidoc:124]: Oops, missed this while updating the docs. Thanks for pointing it out.

**Comment by ShourieG** [comment]
@efd6 addressed all the pr comments and pushed the necessary changes.

File: x-pack/filebeat/input/azureblobstorage/input_test.go [hunk 12]
```diff
@@ -0,0 +1,328 @@
...
+)
+
+func Test_StorageClient(t *testing.T) {
+	tests := []struct {
+		name            string
+		baseConfig      map[string]interface{}
+		mockHandler     func() http.Handler
+		expected        map[string]bool
+		isError         error
+		unexpectedError error
```
Comment (efd6) [inline x-pack/filebeat/input/azureblobstorage/input_test.go:38]: Not for now: I would call these `wantErr` and `waitErr` to be more clearly named. (`isFoo` suggests a boolean and `unexpectedFoo` suggests something that is unexpected).

File: x-pack/filebeat/input/azureblobstorage/input_test.go [hunk 13]
```diff
@@ -0,0 +1,328 @@
...
+						errStr := strings.Join(arr[1:], "\n")
+						assert.Equal(t, tt.isError.Error(), errStr)
+					} else {
+						assert.EqualError(t, g.Wait(), tt.isError.Error())
+					}
+					cancel()
+				} else {
+					assert.NoError(t, g.Wait())
+					cancel()
+				}
```
Comment (efd6) [inline x-pack/filebeat/input/azureblobstorage/input_test.go:290]: Not for now.
```suggestion
				} else {
					assert.NoError(t, g.Wait())
				}
				cancel()
```

**Review by efd6 — APPROVED** [review]
LGTM; suggestions can be left for later.
